doc: usb: add initial USB device configuration howto (#73877)
zephyrproject-rtos/zephyr: doc/connectivity/usb/device_next/usb_device.rst

SoC has multiple controllers, they can be used simultaneously. Full and
high-speed device controllers are supported. It also provides support for
registering multiple function or class instances to a configuration at runtime,
or changing the configuration later. It has built-in support for several USB
classes and provides an API to implement custom USB functions.

The new USB device support is considered experimental and will replace
:ref:`usb_device_stack`.

Built-in functions
==================

The USB device stack has built-in USB functions. Some can be used directly in
the user application through a special API, such as HID or Audio class devices,
while others use a general Zephyr RTOS driver API, such as MSC and CDC class
implementations. The *Identification string* identifies a class or function
instance (n) and is used as an argument to the :c:func:`usbd_register_class`.

+-----------------------------------+-------------------------+-------------------------+
| Class or function                 | User API (if any)       | Identification string   |
+===================================+=========================+=========================+

> | USB Audio 2 class                 | :ref:`uac2_device`      | uac2_(n)                |

Reviewer: you may want to play with / use the following role:

```suggestion
| USB Audio 2 class                 | :ref:`uac2_device`      | :samp:`uac2_{n}`                |
```

even if you don't end up using it, putting the id strings between double backquotes might make things easier to read.

be instantiated. A USB controller device can be assigned to multiple contexts,
but only one context can be initialized and used at a time. Context properties
must not be directly accessed or manipulated by the application.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc device instantiation start
   :end-before: doc device instantiation end

Your USB device may have manufacturer, product, and serial number string
descriptors. To instantiate these string descriptors, the application should
use the appropriate :c:macro:`USBD_DESC_MANUFACTURER_DEFINE`,
:c:macro:`USBD_DESC_PRODUCT_DEFINE`, and
:c:macro:`USBD_DESC_SERIAL_NUMBER_DEFINE` macros. String descriptors also
require a single instantiation of the language descriptor using the
:c:macro:`USBD_DESC_LANG_DEFINE` macro.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc string instantiation start
   :end-before: doc string instantiation end

String descriptors must be added to the device context at runtime before
initializing the USB device with :c:func:`usbd_add_descriptor`.

> .. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
>    :language: c
>    :start-after: doc add string descriptor start
>    :end-before: doc add string descriptor end

Reviewer: here and the other `literalinclude`s as well (even the ones that don't need it, IMO) -- this makes the code left aligned even when the included section happens to be indented.

```suggestion
.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :dedent:
   :start-after: doc add string descriptor start
   :end-before: doc add string descriptor end
```

This will make the below flush left

<img width="837" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/add98898-2ecd-4a2b-99a9-ea471c3fbf59">

:c:macro:`USBD_DEVICE_DEFINE` macro. This creates a static
:c:struct:`usbd_context` variable with a given name. Any number of contexts may
be instantiated. A USB controller device can be assigned to multiple contexts,
but only one context can be initialized and used at a time. Context properties
must not be directly accessed or manipulated by the application.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc device instantiation start
   :end-before: doc device instantiation end

Your USB device can use manufacturer, product, and serial number string
descriptors. The application should use the appropriate
:c:macro:`USBD_DESC_MANUFACTURER_DEFINE`, :c:macro:`USBD_DESC_PRODUCT_DEFINE`,
and :c:macro:`USBD_DESC_SERIAL_NUMBER_DEFINE` macros. Optional string
descriptors also require a single instantiation of the language descriptor
using the :c:macro:`USBD_DESC_LANG_DEFINE` macro.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc string instantiation start
   :end-before: doc string instantiation end

> USB device requires at least one configuration. If the USB controller assigned
> to the device context supports high speed, it needs at least one configuration
> instance per speed. The application should use

Reviewer 2: What about:
```suggestion
USB device requires at least one configuration instance per supported speed.
The application should use
```

:c:macro:`USBD_CONFIGURATION_DEFINE` to instantiate a configuration. Later, USB
device functions are assigned to a configuration.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc configuration instantiation start
   :end-before: doc configuration instantiation end

> Any string descriptor must be added to the device context at runtime before

Reviewer 2: ```suggestion
String descriptors must be added to the device context at runtime before
```

initializing the USB device with :c:func:`usbd_add_descriptor`.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc add string descriptor start
   :end-before: doc add string descriptor end

Each configuration instance for a specific speed must be added to the device
context at runtime before the USB device is initialized using
:c:func:`usbd_add_configuration`. See :c:enumerator:`USBD_SPEED_FS` and

> :c:enumerator:`USBD_SPEED_HS`. For each of the spedific speed configuration

Reviewer 2: ```suggestion
:c:enumerator:`USBD_SPEED_HS`. For each of the specific speed configuration
```

number is assigned based on the order of addition to the context, means the
first full speed configuration will get number one, so called high speed, and

> further accordingly.

Reviewer 2: I got lost at "so called high speed"

Author: Me too, changed it.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc configuration register start
   :end-before: doc configuration register end

Although we have already done a lot, this USB device has no function. A device

> can have multiple configuration with different set of function, event at

Reviewer 2: ```suggestion
can have multiple configurations with different set of functions at different speeds.
```

different speed. A function or class can be registered on a USB device before
it is initialized using :c:func:`usbd_register_class`. The specific
configuration is specified using :c:enumerator:`USBD_SPEED_FS` or
:c:enumerator:`USBD_SPEED_HS` and the configuration number.  For simple cases,
:c:func:`usbd_register_all_classes` can be used to register all available
instances.

.. literalinclude:: ../../../../samples/subsys/usb/common/sample_usbd_init.c
   :language: c
   :start-after: doc functions register start
   :end-before: doc functions register end

The last step in the preparation is to initialize the device with
:c:func:`usbd_init`. After this, the configuration of the device cannot be
changed. A device can be deinitialized with :c:func:`usbd_shutdown` and all
instances can be reused, but the previous steps must be repeated. So it is
possible to shutdown a device, register another type of configuration or
function, and initialize it again.  At the USB controller level,
:c:func:`usbd_init` does only what is necessary to detect VBUS changes. There
are controller types where the next step is only possible if a VBUS signal is

> present, more on this later.

Reviewer 2: ```suggestion
present.
```

The "more on this later" only adds confusion and seem completely unnecessary here.
